### PR TITLE
Add stats on rescued/ignored tasks

### LINKED ISSUE #14393: Task state rescued
##### Issue Type:
- Feature Idea
##### Component Name:
- Feature core
##### Ansible Version:
- 2.0.0.2
##### Ansible Configuration:
- No relevant changes in Ansible configuration for this feature idea.
##### Environment:
- N/A (Archlinux, Ansible installed via pip in virtualenv, python 2.7)
##### Summary:

`blocks` feature with `rescue` results in play recap with failed events.

I'd pretty much like to see another event like `rescued`: I know, there was an error, but this was possibly on purpose and got dealt with in a rescue block and it is therefor not an error in terms of a plays recap.
##### Steps To Reproduce:

```
# /tmp/playbook.yml

---
- name: Produce failed event in rescue operations
  hosts: localhost
  connection: local
  tasks:
    - include: block-rescue.yml
```

```
# /tmp/block-rescue.yml

---
- block:
  - name: Create failed event to rescue
    fail:
      msg: I fail on purpose
  rescue:
  - name: I am now in rescue mode
    debug:
      msg: "I am rescuing the failed event"
```

```
ansible-playbook /tmp/playbook.yml
```
##### Expected Results:

No failed events in play recap (reporting).
##### Actual Results:

There is one failed event in play recap (reporting).
##### Further observation

The following playbook (the same as above without using include) removes the failed event from the recap completely.

```

---
- hosts: localhost
  connection: local
  gather_facts: false
  tasks:
    - block:
      - name: Create failed event to rescue
        fail:
          msg: I fail on purpose
      rescue:
      - name: I am now in rescue mode
        debug:
          msg: "I am rescuing the failed event"
```

This feels quite inconsistent.

### LINKED ISSUE #20346: Report failed tasks that were rescued separately from other failures
##### ISSUE TYPE
 - Feature Request

##### COMPONENT NAME
Ansible core

##### ANSIBLE VERSION
```
ansible 2.2.1.0
  config file = 
  configured module search path = Default w/o overrides
```

##### CONFIGURATION
ANSIBLE_GATHERING=smart
ANSIBLE_HASH_BEHAVIOUR=merge
ANSIBLE_REMOTE_TEMP=/tmp
ANSIBLE_RETRY_FILES_ENABLED=no
ANSIBLE_SSH_PIPELINING=yes
ANSIBLE_TRANSPORT=local

##### OS / ENVIRONMENT
CentOS 7, with ansible running in local mode.

##### SUMMARY
Under Ansible 2.1, a block that had a failed task, but that was rescued, did not cause Ansible to report any failures. In other words, at the end of the play, I would see "failed=0".

With Ansible 2.2.1, even though a failed task is rescued, Ansible reports "failed=1" at the end.

##### STEPS TO REPRODUCE
Create a block..rescue sequence, containing a task that fails, and run the playbook. Ansible will report failed tasks at the end.

```yaml
- hosts: localhost
  gather_facts: no
  tasks:
  - block:
    - fail: msg="This failure will be rescued"
    rescue:
    - debug: msg=NOOP
```

##### EXPECTED RESULTS
The summary line at the end should have shown failed=0

##### ACTUAL RESULTS
The summary line shows failed=1

```
PLAY RECAP *********************************************************************
localhost                  : ok=1    changed=0    unreachable=0    failed=1
```

### LINKED ISSUE #24525: Feature Idea: Change font color of failed tasks that are "rescued" to bright purple (color for warn) or some non-red color
##### ISSUE TYPE

 - Feature Idea

##### ANSIBLE VERSION

```
ansible 2.3.0.0
  configured module search path = [u'modules']
  python version = 2.7.13 (default, Apr 13 2017, 16:18:01) [GCC 4.2.1 Compatible Apple LLVM 8.1.0 (clang-802.0.38)]
```

##### SUMMARY

When scanning logs, people are confused when tasks that are "rescued" are shown in red thinking it's a failure. In reality the error was handled or the error actually came later. It's not really a big deal in terms of functionality, mainly usability, just easier to search for errors quickly and to not confuse people unfamiliar with ansible. Changing all error tasks to non-red would probably not be helpful.

##### STEPS TO REPRODUCE

```yaml
---

- hosts: localhost
  connection: local
  tasks:
    - block:
      - name: i'm red
        command: "false"
      rescue:
        - name: i'm yellow
          command: "true"
    - name: still going
      command: "true"
```
image

##### EXPECTED RESULTS

Non-red color for rescued tasks.

##### ACTUAL RESULTS

All failed tasks are shown in red.

```

```

### LINKED ISSUE #31245: Better stats/status on rescue/ignored tasks
##### SUMMARY
Expand the number of stats collected during task execution beyond simply changed/failed/unreachable. For example, add in a new stat for rescued tasks (which are failed but also have rescues running).

##### ISSUE TYPE
- Feature Idea

##### COMPONENT NAME
core

### PR DESCRIPTION
**Pull request by mkrizek**
##### SUMMARY
PLEASE PING THE TOWER TEAM BEFORE MERGING

Following up on the work that @jimi-c has done in his `improved_stats` branch https://github.com/jimi-c/ansible/commit/2832d082b2840835654667d9598bb1dff3ecd8b9.

This PR adds two counters to `PLAY RECAP`:
* `ignored` for tasks that failed but had set `ignore_errors: yes`,
* `rescued` for tasks that failed initially but the rescue section was executed; note that `rescued` tasks are no longer counted as `failed` as before this PR

Fixes https://github.com/ansible/ansible/issues/31245
Fixes https://github.com/ansible/ansible/issues/20346
Fixes #24525
Fixes #14393

##### ISSUE TYPE

- Feature Pull Request

##### COMPONENT NAME

callbacks
##### ANSIBLE VERSION

```paste below
2.8
```

### CODE REVIEW

**Comment by mkrizek** [comment]
cc @AlanCoding

File: test/integration/targets/callback_default/callback_default.out.hide_skipped_ok.stdout
Comment (bcoca) [inline test/integration/targets/callback_default/callback_default.out.hide_skipped_ok.stdout]: we might want to change to `failed=Y (ignored=X)` so the 'sum of tasks' seems to stay the same, these are not at the same level as the general counters, they are subsets of the existing categories

File: test/integration/targets/callback_default/callback_default.out.default.stdout
Comment (bcoca) [inline test/integration/targets/callback_default/callback_default.out.default.stdout]: why are we loosing `skipped`?

File: lib/ansible/plugins/callback/default.py
Comment (bcoca) [inline lib/ansible/plugins/callback/default.py:331]: I was thinking that we should show them as subgroups, otherwise task numbers don't match up:  
```
 u"%s : %s(%s) %s %s %s(%s) %s " %``
...
colorize(u'ok', t['ok'], C.COLOR_OK), 
colorize(u'rescued', t['rescued'], C.COLOR_OK),
...
 colorize(u'failed', t['failures'], C.COLOR_ERROR),
colorize(u'ignored', t['ignored'], C.COLOR_WARN),
```

Comment (mkrizek) [reply lib/ansible/plugins/callback/default.py:331]: But `rescued` tasks are not counted as `ok` so unsure `ok=X (rescued=Y)` makes sense? Not sure `failed=X (ignored=Y)` makes sense either since those two are mutually exclusive, you can have either one or the other.

See this playbook and its recap:
```
- hosts: localhost
  gather_facts: no
  tasks:
    - name: ok1
      debug:

    - name: ok2 and changed
      debug:
      changed_when: True

    - name: skipped
      debug:
      when: False

    - block:
        - name: rescued
          fail:
      rescue:
        - name: ok3
          debug:

    - name: ok4 and ignored
      fail:
      ignore_errors: yes

    - name: failed
      fail:
```

`localhost                  : ok=4    changed=1    unreachable=0    failed=1    skipped=1    rescued=1    ignored=1`

Task numbers don't match up because `changed` and `ignored` are counted as `ok` as well.

File: docs/docsite/rst/porting_guides/porting_guide_2.8.rst
Comment (acozine) [inline docs/docsite/rst/porting_guides/porting_guide_2.8.rst]: I'd start with the part users will see, then describe how it works:

"Play recap now counts `ignored` and `rescued` tasks as well as `ok`, `changed`, `skipped` and `failed` tasks, thanks to two additional stat counters in the `default` callback plugin. Tasks that fail and have `ignore_errors: yes` set are listed as `ignored`. Tasks that fail and then execute a rescue section are listed as `rescued`. Note that `ignored` tasks are also included in the total count of `failed` tasks, but `rescued` tasks are not."

Comment (mkrizek) [reply docs/docsite/rst/porting_guides/porting_guide_2.8.rst]: Thanks for re-wording. Couple of notes though:
* "`ok`, `changed`, `skipped` and `failed` tasks" - if we're listing all currently existing counters, should we mention `unreachable` too?
* "Note that ignored tasks are also included in the total count of failed tasks" - this isn't true as `ignored` and `failed` are mutually exclusive. So I'd leave the last sentence as currently is in the PR (or re-word it) since it clarifies what has changed from previous versions?

File: lib/ansible/plugins/callback/default.py
Comment (bcoca) [reply lib/ansible/plugins/callback/default.py:331]: I seem to have confused the accounting, so `ok (ignored)` .. i assumed rescued was 'failed but rescued' but it seems to account for 'rescue tasks run'?

Comment (mkrizek) [reply lib/ansible/plugins/callback/default.py:331]: If we do `ok (ignore)`, should we do also `ok (changed)`, so `ok=X (changed=Y, ignored=Z)` then? Because task number hasn't been matched up ever as `changed` tasks are counted as `ok` too.

No, `rescued` are 'failed but rescued' tasks, but they are NOT counted as `failed` so in the example above it's
```
        - name: rescued
          fail:
```
so `rescued=1` (and `failed=0` if it was just this block).

Comment (bcoca) [reply lib/ansible/plugins/callback/default.py:331]: yeah, i think that will alleviate a lot of confusion with the status of tasks

**Comment by jladdjr** [comment]
@jctanner - Looks like this is ready for review (seeing `core_review` label). Know that @bcoca is going to be out for a bit, though. Is there anyone else on the core team who can review the work that's been done here? Tower has a 3.5 feature that wraps around this change (see https://github.com/ansible/awx/issues/466).

**Comment by jctanner** [comment]
I don't think this is waiting on bcoca anymore, as all his review comments have been answered.

**Comment by Akasurde** [comment]
I don't want to block as such but let me know if we can address this two issue - 

1. https://github.com/ansible/ansible/issues/24525
2. https://github.com/ansible/ansible/issues/14393

**Review by jborean93 — APPROVED** [review]
The changes look good to me, testing them locally do what they should be doing. I'm not sure if we have concerns around changing the entries in the recap output but it sounds like that conversation has already been had

**Comment by jborean93** [comment]
Thanks @Akasurde for sharing those issues. I've commented on them as this PR will fix their issues.

**Review by jladdjr — APPROVED** [review]

**Comment by jborean93** [comment]
rebuild_merge

**Pull request merged**
